zk-token-sdk/src/instruction/transfer/encryption.rs

```diff
@@ -2,48 +2,49 @@
 use crate::{
     encryption::{
         elgamal::{DecryptHandle, ElGamalPubkey},
+        grouped_elgamal::{GroupedElGamal, GroupedElGamalCiphertext},
         pedersen::{Pedersen, PedersenCommitment, PedersenOpening},
     },
     zk_token_elgamal::pod,
 };
 
-// TransferAmountEncryption
-#[derive(Clone)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
 #[repr(C)]
 #[cfg(not(target_os = "solana"))]
-pub struct TransferAmountEncryption {
-    pub commitment: PedersenCommitment,
-    pub source_handle: DecryptHandle,
-    pub destination_handle: DecryptHandle,
-    pub auditor_handle: DecryptHandle,
-}
+pub struct TransferAmountCiphertext(pub(crate) GroupedElGamalCiphertext<3>);
 
 #[cfg(not(target_os = "solana"))]
-impl TransferAmountEncryption {
+impl TransferAmountCiphertext {
     pub fn new(
         amount: u64,
         source_pubkey: &ElGamalPubkey,
         destination_pubkey: &ElGamalPubkey,
         auditor_pubkey: &ElGamalPubkey,
     ) -> (Self, PedersenOpening) {
-        let (commitment, opening) = Pedersen::new(amount);
-        let transfer_amount_encryption = Self {
-            commitment,
-            source_handle: source_pubkey.decrypt_handle(&opening),
-            destination_handle: destination_pubkey.decrypt_handle(&opening),
-            auditor_handle: auditor_pubkey.decrypt_handle(&opening),
-        };
+        let opening = PedersenOpening::new_rand();
+        let grouped_ciphertext = GroupedElGamal::<3>::encrypt_with(
+            [source_pubkey, destination_pubkey, auditor_pubkey],
+            amount,
+            &opening,
+        );
 
-        (transfer_amount_encryption, opening)
+        (Self(grouped_ciphertext), opening)
     }
 
-    pub fn to_pod(&self) -> pod::TransferAmountEncryption {
-        pod::TransferAmountEncryption {
-            commitment: self.commitment.into(),
-            source_handle: self.source_handle.into(),
-            destination_handle: self.destination_handle.into(),
-            auditor_handle: self.auditor_handle.into(),
-        }
+    pub fn get_commitment(&self) -> &PedersenCommitment {
+        &self.0.commitment
+    }
+
+    pub fn get_source_handle(&self) -> &DecryptHandle {
+        self.0.handles.get(0).unwrap()
+    }
+
+    pub fn get_destination_handle(&self) -> &DecryptHandle {
+        self.0.handles.get(1).unwrap()
+    }
+
+    pub fn get_auditor_handle(&self) -> &DecryptHandle {
+        self.0.handles.get(2).unwrap()
```

Comment:
Could you add comments or expect messages that make it clear why it's safe to unwrap these?

Reply:
Yep, that is a good idea!